bip-musig2.mediawiki

```diff
@@ -410,6 +391,49 @@ The reference implementation is for demonstration purposes only and not to be us
 
 == Remarks on Security and Correctness ==
 
+=== Modifications to Nonce Generation ===
+
+Implementors must avoid modifying the ''NonceGen'' algorithm without being fully aware of the implications.
+The following two modifications are secure when applied correctly and may be useful in special circumstances, e.g., in very restricted environments where secure randomness is not available.
+
+First, on systems where obtaining uniformly random values is much harder than maintaining a global atomic counter, it can be beneficial to modify ''NonceGen''.
+Instead of drawing ''rand' '' uniformly at random, ''rand' '' can be the value of an atomic counter that is incremented whenever it is read.
+With this modification, the secret signing key ''sk'' of the signer generating the nonce is '''not''' an optional argument and must be provided to ''NonceGen''.
+The security of the resulting scheme is then depending on the requirement that the counter must never return the same output in two ''NonceGen'' invocations with the same ''sk''.
+
+Second, if there is a unique signer who is supposed to send the ''pubnonce'' last, it is possible to modify nonce generation for this single signer to not require high-quality randomness.
+Such a nonce generation algorithm ''DeterministicSign'' is specified below.
+Note that there are no optional arguments except ''rand'', which can be ommitted if randomness is entirely unavailable.
+''DeterministicSign'' requires the argument ''othernonce'' which is the output of ''NonceAgg'' run on the ''pubnonce'' value of all other signers.
+Hence, using ''DeterministicSign'' is only possible for the last signer to generate a nonce and makes the signer stateless, similar to the stateless signer described in the [[#nonce-generation|Nonce Generation]] section.
```

Comment:
Right, it's very similar to the other "stateless" signer. 

I believe the following is true: If you want to make one signer stateless, it can't hurt to make that one deterministic.

So I suggest we unify this and only offer a single way to make a signer stateless, which is the deterministic one. I think an easy way to implement this in the document is to refer directly to the deterministic nonce gen in the "Nonce Generation" section.

Reply:
It hurts in the sense that it requires an implementation of a new algorithm `DeterministicSign` but the same can be achieved in a simpler way, by reusing the existing algorithms.

Reply:
Yeah, that's true. On the other hand, using `DeterministicSign` should be much safer the single signer, so I'm not sure if still worth to keep the added complexity of two variants. I think we should at least say that using `DeterministicSign` is highly recommend when a signer is made stateless.